File: src/main/java/me/deecaad/weaponmechanicsplus/WeaponMechanicsPlus.java
package me.deecaad.weaponmechanicsplus;

import me.deecaad.core.events.QueueSerializerEvent;
import me.deecaad.core.file.SerializerInstancer;
import me.deecaad.core.file.TaskChain;
import me.deecaad.core.utils.Debugger;
import me.deecaad.core.utils.FileUtil;
import me.deecaad.core.utils.LogLevel;
import me.deecaad.weaponmechanics.WeaponMechanics;
import me.deecaad.weaponmechanics.lib.auto.UpdateChecker;
import me.deecaad.weaponmechanics.lib.auto.UpdateInfo;
import me.deecaad.weaponmechanics.lib.bstats.bukkit.Metrics;
import me.deecaad.weaponmechanics.weapon.WeaponHandler;
import me.deecaad.weaponmechanics.weapon.projectile.ProjectilesRunnable;
import me.deecaad.weaponmechanicsplus.weapon.firemode.FireModeTriggerListener;
import me.deecaad.weaponmechanicsplus.weapon.listeners.AttachmentListeners;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.plugin.Plugin;

import java.io.File;
import java.io.IOException;
import java.util.jar.JarFile;
import java.util.logging.Logger;

public class WeaponMechanicsPlus {

    private static WeaponMechanicsPlus plugin;

    private WeaponMechanicsPlusLoader javaPlugin;
    private UpdateChecker update;
    private Metrics metrics;
    private Debugger debug;

    WeaponMechanicsPlus(WeaponMechanicsPlusLoader javaPlugin) {
        this.javaPlugin = javaPlugin;

        plugin = this;
    }

    public void onLoad() {
        int level = getConfig().getInt("Debug_Level", 2);
        boolean printTraces = getConfig().getBoolean("Print_Traces", false);
        debug = new Debugger(getLogger(), level, printTraces);
    }

    public void onEnable() {
        writeFiles();
        registerDebugger();
        registerUpdateChecker();
        registerBStats();
        registerSerializerQueue();
    }

    public void onDisable() {
    }

    public FileConfiguration getConfig() {
        return javaPlugin.getConfig();
    }

    public Logger getLogger() {
        return javaPlugin.getLogger();
    }

    public File getDataFolder() {
        return javaPlugin.getDataFolder();
    }

    public ClassLoader getClassLoader() {
        return javaPlugin.getClassLoader0();
    }

    public File getFile() {
        return javaPlugin.getFile0();
    }

    public static Debugger getDebug() {
        return plugin.debug;
    }

    private void writeFiles() {
        // Create files
        if (!getDataFolder().exists() || getDataFolder().listFiles() == null || getDataFolder().listFiles().length == 0) {
            debug.info("Copying files from jar (This process may take up to 30 seconds during the first load!)");
            FileUtil.copyResourcesTo(getClassLoader().getResource("WeaponMechanicsPlus"), getDataFolder().toPath());
        }
    }

    private void registerDebugger() {
        debug.permission = "weaponmechanicsplus.errorlog";
        debug.msg = "WeaponMechanicsPlus had %s error(s) in console.";
        debug.start(javaPlugin);
    }

    private void registerUpdateChecker() {

        // TODO WHEN APPROVED CHANGE ID
        boolean todoWaiting = true;
        if (todoWaiting) return;

        if (!getConfig().getBoolean("Update_Checker.Enabled", true)) return;

        update = new UpdateChecker(javaPlugin, UpdateChecker.spigot(1, "WeaponMechanicsPlus"));
        Listener listener = new Listener() {
            @EventHandler
            public void onJoin(PlayerJoinEvent event) {
                if (event.getPlayer().isOp()) {
                    new TaskChain(javaPlugin)
                            .thenRunAsync((callback) -> update.hasUpdate())
                            .thenRunSync((callback) -> {
                                UpdateInfo update = (UpdateInfo) callback;
                                if (callback != null)
                                    event.getPlayer().sendMessage(ChatColor.RED + "WeaponMechanicsPlus is out of date! " + update.current + " -> " + update.newest);

                                return null;
                            });
                }
            }
        };

        Bukkit.getPluginManager().registerEvents(listener, javaPlugin);
    }

    private void registerBStats() {
        if (metrics != null) return;

        debug.debug("Registering bStats");

        // See https://bstats.org/plugin/bukkit/WeaponMechanicsPlus/16382. This is
        // the bStats plugin id used to track information.
        int id = 16382;

        this.metrics = new Metrics(javaPlugin, id);
    }

    private void registerSerializerQueue() {
        Listener listener = new Listener() {
            @EventHandler
            public void onJoin(QueueSerializerEvent event) {
                // Perfect place to register all things to WM ;D
                if (!event.getSourceName().equals("WeaponMechanics")) return;

                // Register serializers
                try {
                    event.addSerializers(new SerializerInstancer(new JarFile(getFile())).createAllInstances(getClassLoader()));
                } catch (IOException e) {
                    debug.log(LogLevel.WARN, "Failed to add serializers...", e);
                }

                // Register trigger listeners
                WeaponHandler weaponHandler = WeaponMechanics.getWeaponHandler();
                weaponHandler.addTriggerListener(new FireModeTriggerListener());

                // Register projectile script manager
                ProjectilesRunnable projectilesRunnable = WeaponMechanics.getProjectilesRunnable();
                projectilesRunnable.addScriptManager(new PlusScriptManager(javaPlugin));
            }
        };

        Bukkit.getPluginManager().registerEvents(listener, javaPlugin);
        Bukkit.getPluginManager().registerEvents(new AttachmentListeners(), javaPlugin);
    }

    public static Plugin getPlugin() {
        return plugin.javaPlugin;
    }
}

File: src/main/kotlin/me/deecaad/weaponmechanicsplus/WeaponMechanicsPlus.kt
package me.deecaad.weaponmechanicsplus

import me.deecaad.core.events.QueueSerializerEvent
import me.deecaad.core.file.SerializerInstancer
import me.deecaad.core.file.TaskChain
import me.deecaad.core.utils.Debugger
import me.deecaad.core.utils.FileUtil
import me.deecaad.core.utils.LogLevel
import me.deecaad.weaponmechanics.WeaponMechanics
import me.deecaad.weaponmechanics.lib.auto.UpdateChecker
import me.deecaad.weaponmechanics.lib.auto.UpdateInfo
import me.deecaad.weaponmechanics.lib.bstats.bukkit.Metrics
import me.deecaad.weaponmechanicsplus.weapon.firemode.FireModeTriggerListener
import me.deecaad.weaponmechanicsplus.weapon.listeners.AttachmentListeners
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.plugin.Plugin
import java.io.File
import java.io.IOException
import java.util.jar.JarFile
import java.util.logging.Logger

class WeaponMechanicsPlus internal constructor(private val javaPlugin: WeaponMechanicsPlusLoader) {

    private lateinit var update: UpdateChecker
    private lateinit var metrics: Metrics
    private lateinit var debug: Debugger

    val config: FileConfiguration
        get() = javaPlugin.config
    val logger: Logger
        get() = javaPlugin.logger
    val dataFolder: File
        get() = javaPlugin.dataFolder
    val classLoader: ClassLoader
        get() = javaPlugin.classLoader0
    val file: File
        get() = javaPlugin.file0

    init {
        plugin = this
    }

    fun onLoad() {
        val level = config.getInt("Debug_Level", 2)
        val printTraces = config.getBoolean("Print_Traces", false)
        debug = Debugger(logger, level, printTraces)
    }

    fun onEnable() {
        writeFiles()
        registerDebugger()
        registerUpdateChecker()
        registerBStats()
        registerSerializerQueue()
    }

    fun onDisable() {}

    private fun writeFiles() {
        // Create files
        if (!dataFolder.exists() || dataFolder.listFiles() == null || dataFolder.listFiles()?.size == 0) {
            debug.info("Copying files from jar (This process may take up to 30 seconds during the first load!)")
            FileUtil.copyResourcesTo(classLoader.getResource("WeaponMechanicsPlus"), dataFolder.toPath())
        }
    }

    private fun registerDebugger() {
        debug.permission = "weaponmechanicsplus.errorlog"
        debug.msg = "WeaponMechanicsPlus had %s error(s) in console."
        debug.start(javaPlugin)
    }

    private fun registerUpdateChecker() {

        // TODO WHEN APPROVED CHANGE ID
        val todoWaiting = true
        if (todoWaiting) return
        if (!config.getBoolean("Update_Checker.Enabled", true)) return
        update = UpdateChecker(javaPlugin, UpdateChecker.spigot(1, "WeaponMechanicsPlus"))
        val listener: Listener = object : Listener {
            @EventHandler
            fun onJoin(event: PlayerJoinEvent) {
                if (event.player.isOp) {
                    TaskChain(javaPlugin)
                        .thenRunAsync { callback: Any? -> update!!.hasUpdate() }
                        .thenRunSync { callback: Any? ->
                            val update = callback as UpdateInfo?
                            if (callback != null) event.player.sendMessage(ChatColor.RED.toString() + "WeaponMechanicsPlus is out of date! " + update!!.current + " -> " + update.newest)
                            null
                        }
                }
            }
        }
        Bukkit.getPluginManager().registerEvents(listener, javaPlugin)
    }

    private fun registerBStats() {
        if (metrics != null) return
        debug!!.debug("Registering bStats")

        // See https://bstats.org/plugin/bukkit/WeaponMechanicsPlus/16382. This is
        // the bStats plugin id used to track information.
        val id = 16382
        metrics = Metrics(javaPlugin, id)
    }

    private fun registerSerializerQueue() {
        val listener = object : Listener {
            @EventHandler
            fun onJoin(event: QueueSerializerEvent) {
                // Perfect place to register all things to WM ;D
                if (event.sourceName != "WeaponMechanics") return

                // Register serializers
                try {
                    event.addSerializers(SerializerInstancer(JarFile(file)).createAllInstances(classLoader))
                } catch (e: IOException) {
                    debug!!.log(LogLevel.WARN, "Failed to add serializers...", e)
                }

                // Register trigger listeners
                val weaponHandler = WeaponMechanics.getWeaponHandler()
                weaponHandler.addTriggerListener(FireModeTriggerListener())

                // Register projectile script manager
                val projectilesRunnable = WeaponMechanics.getProjectilesRunnable()
                projectilesRunnable.addScriptManager(PlusScriptManager(javaPlugin))
            }
        }
        Bukkit.getPluginManager().registerEvents(listener, javaPlugin)
        Bukkit.getPluginManager().registerEvents(AttachmentListeners(), javaPlugin)
    }

    companion object {
        private lateinit var plugin: WeaponMechanicsPlus
        fun getDebug(): Debugger? {
            return plugin.debug
        }

        fun getPlugin(): Plugin {
            return plugin.javaPlugin
        }
    }
}